### PR TITLE
Customise AppVeyor Test 'passed', 'failed', 'skipped' status

### PR DESCRIPTION
**Pull request by NZSmartie**
This is to allow users to display text in their own language, or (in my case) unicode characters. For example:
```
http://img.shields.io/appveyor/tests/NZSmartie/coap-net-iu0to.svg?passed=%E2%9C%94&failed=%E2%9D%8C&skipped=%E2%9D%97
```
would show the following badge
image

### CODE REVIEW

**Comment by tooomm** [comment]
Interesting idea, I would like to see a documentation for that in the `parameters` section!

**Comment by NZSmartie** [comment]
One of the problems is that it's difficult to document what a badge has. There has been progress for improving registering services API (#963) which could allow for standarised documentation for badges. 

The query parameters introduced in this PR are specific to the AppVeyor test status badge. Thus, adding information to the parameters section at the bottom would be very misleading since it does not apply to other badges.

**Comment by tooomm** [comment]
Ohh, I thought you could customize any badge in the sense that it would replace given words with the selected characters!

But undocumented or poorly documented features are always bad.
There could be a hint to this pr (or a wiki article) next to the appveyor badge at least for example.
The best solution might be to have an explanation right within the appveyor badge pop-up, like the text that all GitHub badges have. Check this screenshot: https://github.com/badges/shields/issues/1311#issuecomment-348707190
The `Documenation` section is only there for "GitHub badges", you could use that to explain the appveyor specific custom options.

A general availability of something like this would be cool, sure.

**Comment by NZSmartie** [comment]
Heh, that's new from last time I did a PR. will add what I can. cheers @tooomm

**Comment by paulmelnikow** [comment]
Another way to document it would be an additional badge example. This also needs a test.

Could you clarify why this is better than the standard text? Since a goal of the project is to provide consistent badges, we try to choose a good default and avoid providing customization options. Do you think we should change the standard text? Translation is one thing (all the badges are in English; only some of that can be overridden) but what you're wanting here is more like an alternate format.

**Comment by platan** [comment]
Code and documentation looks good fore me. Additionally you could add some test for overriding values (default values are already tested in service tests). 
But badges (examples from documentation) does not look good at my PC (Ubuntu 16.04)
Chrome:
chrome-screenshot-jsbin com-2017-12-04-19-57-15-495
Firefox:
firefox-screenshot-2017-12-4 js bin

**Comment by NZSmartie** [comment]
Regarding the mangled badges in development builds: Grab a copy of `Verdana.ttf` and place it in the repository root directory. then it'll look right

**Comment by platan** [comment]
@NZSmartie with the font it works great! Thanks!
fixed-screenshot-jsbin com-2017-12-04-22-07-53-029

**Comment by NZSmartie** [comment]
@paulmelnikow This PR is an alternative way of displaying the AppVeyor test badge. Because other components of the badge can already be customised (`label`, `style`, `link`, `logo`, `logoWidth`, `link`, `colorA` and `colorB`), adding badge specific customisation seemed like a not too far stretch. 

The main motivation is that badges can get very wide when all 3 metric labels (passed, failed and skipped) are displayed. Allowing use of single unicode helps free up screen space

**Comment by paulmelnikow** [comment]
> The main motivation is that badges can get very wide when all 3 metric labels (passed, failed and skipped) are displayed. Allowing use of single unicode helps free up screen space

That makes sense to me.

Another way we could go about this is to add a flag like `?valueStyle=compact` that uses compact values without needing to specify unicode characters. That helps our goal of uniformity, while also being helpful for another use case.

What do you think about this tweak to the format?



https://img.shields.io/badge/build-%E2%9C%94%2012%20%20%E2%9D%8C%203%20%20%E2%9D%97%203-orange.svg

I'm not sure if the **!** is a clear way to convey skipped – it seems more like errored. I wonder if there is another symbol that might be more understandable.

The **x** and **!** characters don't look great on my machine because they display in red:

screen shot 2017-12-04 at 6 33 38 pm

**Comment by paulmelnikow** [comment]
Another possibility:



https://img.shields.io/badge/tests-%E2%9C%94%2012%20%20%E2%9C%98%203%20%20%E2%A4%B8%203-orange.svg

File: service-tests/appveyor.js [hunk 1]
```diff
@@ -35,6 +39,16 @@ t.create('tests status on master branch')
...
 
+// Test AppVeyor tests status badge with custom label
+t.create('tests status with custom labels')
+.get('/tests/NZSmartie/coap-net-iu0to.json?passed=good&failed=bad&skipped=n/a')
+.expectJSONTypes(Joi.object().keys({ name: 'tests', value: isCustomAppveyorTestTotals }));
+
+// Test AppVeyor tests status badge with emoji
+t.create('tests status with emoji labels')
+.get('/tests/NZSmartie/coap-net-iu0to.json?passed=%E2%9C%94&failed=%E2%9D%8C&skipped=%E2%9D%97')
+.expectJSONTypes(Joi.object().keys({ name: 'tests', value: isEmojiAppveyorTestTotals }));
```
Comment (paulmelnikow) [inline service-tests/appveyor.js]: Tests look great!

**Comment by NZSmartie** [comment]
We could use the info unicode character
https://img.shields.io/badge/tests-%E2%9C%94%2012%20%20%E2%9C%98%203%20%20%E2%84%B9%203-orange.svg
image

**Comment by NZSmartie** [comment]
I like the idea of a `&valueStyle=compact` query parameter. lemme implement that

**Comment by paulmelnikow** [comment]
Does the [information symbol](https://www.fileformat.info/info/unicode/char/2139/index.htm) convey skipped, though?

Here are some other ideas:

1. 
2. 
3. 
4. 
5. 
6. 

I'm rather partial to the last, which is the symbol for "escape."

The "skip track" icon is clear in meaning, but it doesn't look good.

screen shot 2017-12-04 at 9 06 32 pm

**Comment by NZSmartie** [comment]
Hmm, what about tilde? 
example

**Comment by RedSparr0w** [comment]
I quite like the tilde sign, as its pretty simple and gets the point across.

**Comment by paulmelnikow** [comment]
The tilde is simple looking, though how does it convey "skipped"?

**Comment by NZSmartie** [comment]
So, skipped tests may also indicate inconclusive tests (not enough input to assert on). With that in mind. I think a '?' could be used to say that a test isn't failed or passed.

**Comment by NZSmartie** [comment]
If someone want's to use another symbol for skipped tests, they can use the `&skipped=` query parameter 😛

**Comment by platan** [comment]
In my opinion `?` and `~` looks good, however `?` looks a bit like 'unknown' for me.

**Comment by NZSmartie** [comment]
well, skipped tests are unknown, because tests are meant to be assertive. if a test is unknown, then it still needs to be addressed

**Comment by chris48s** [comment]
'skipped' seems to be one of those unfortunate things where every platform uses a different icon. e.g:

screenshot at 2017-12-05 21 40 22

m89mt

19-skipped-tests

In command line utilities, a skipped test is often just presented as a letter `s`

**Comment by NZSmartie** [comment]
*dies inside*

**Comment by paulmelnikow** [comment]
That CLI screenshot that uses the empty circle ain't bad.



https://img.shields.io/badge/tests-%E2%9C%94%2012%20%20%E2%9C%98%203%20%20%E2%97%8B%203-orange.svg

GitHub uses a filled circle, which is similar, though it means "pending."

screen shot 2017-12-05 at 11 43 33 am

**Comment by NZSmartie** [comment]
Filled Circle?
example
example

**Comment by NZSmartie** [comment]
Was asking as the empty circle looks out of place on Windows 10/Firefox
image

File: service-tests/appveyor.js [hunk 2]
```diff
@@ -35,6 +39,21 @@ t.create('tests status on master branch')
   .get('/tests/NZSmartie/coap-net-iu0to/master.json')
   .expectJSONTypes(Joi.object().keys({ name: 'tests', value: isAppveyorTestTotals }));
 
+// Test AppVeyor tests status badge with custom label
+t.create('tests status with custom labels')
+.get('/tests/NZSmartie/coap-net-iu0to.json?passed=good&failed=bad&skipped=n/a')
+.expectJSONTypes(Joi.object().keys({ name: 'tests', value: isCustomAppveyorTestTotals }));
```
[the remark this answers is not in the document]
Comment (NZSmartie) [reply service-tests/appveyor.js]: Would you be willing to let me keep it so that I may use Emoji for the less-serious/more-fun projects? 😛

**Comment by NZSmartie** [comment]
> @paulmelnikow 
> Can we leave off passed, failed, and skipped and just support compact? While translation could be a use case it is not something we support right now, and if we did, we'd let the user specify a language rather than individual strings.

Would you be willing to let me keep it so that I may use Emoji for the less-serious/more-fun projects? 😛

**Comment by tooomm** [comment]
Did you consider this for `skipped` too?
- jumping or leaping arrow: ↷ (`U+21B7`)
- fast-forward arrow: ⏩ (`U+23E9`) (not the ones with the stop sign)
I guess it's the font on Firefox or GitHub page that makes them appear as a button, normally it should look like this:
fast-forward


Is it possible to adjust the color of unicode characters by the way?

**Comment by NZSmartie** [comment]
@tooomm 
[example](https://img.shields.io/badge/tests-%E2%9C%94%2012%20%20%E2%9C%98%203%20%20%E2%86%B7%203-orange.svg)
[example](https://img.shields.io/badge/tests-%E2%9C%94%2012%20%20%E2%9C%98%203%20%20%E2%8F%A9%203-orange.svg)

Hmm the skip emoji (second example) could work

**Comment by tooomm** [comment]
I like the look of `fast-forward`! First one is a bit thin and could be hard to read, yes...

What I generally dislike though is that the connection between icon and count is a bit difficult.
It's hard to match the 3 in the middle to the left or right symbol, or the x to the 12 or 3 just by a quick look.
The spacings are all identical. Could you try putting "()" or "[]" brackets around each pair or separate them with a dash "-" or slash "/" or vertical line "|". Or maybe even just an additional space. The filled circle you used above could work as separator as well!
Check [your example in the starting post](https://github.com/badges/shields/pull/1321#issue-278858854) again, it's easier to understand with the comma for examples.

**Comment by RedSparr0w** [comment]
The emojis show up strange on iOS, it adds an extra square after the emoji
[image1.PNG]

Sent from my iPhone

On 6/12/2017, at 12:36 PM, Roman Vaughan <notifications@github.com<mailto:notifications@github.com>> wrote:


@tooomm<https://github.com/tooomm>
[example]<https://img.shields.io/badge/tests-%E2%9C%94%2012%20%20%E2%9C%98%203%20%20%E2%86%B7%203-orange.svg>
[example]<https://img.shields.io/badge/tests-%E2%9C%94%2012%20%20%E2%9C%98%203%20%20%E2%8F%A9%203-orange.svg>

Hmm the skip emoji (second example) could work

**Comment by RedSparr0w** [comment]
Couldn't add an image to my last post..
image

**Comment by paulmelnikow** [comment]
> I guess it's the font on Firefox or GitHub page that makes them appear as a button

This happens on OS X too:

fastforward

In the arrows category, the downward curving arrow indicates might convey skipping better than the jumping arrow:



Agree that more spacing between would be nice.

I think the spacing is coming out weird in part because of inaccurate text width computation on the server. There's also a behavior I'm noticing which I think is a bug, which is that subsequent spaces get rendered as a single space. In a way, these might be the same problem: when a string has multiple spaces, there's a discrepancy between the way the text width is computed and the way the SVG template renders them.

One of these badges should put lots of spaces between the first and second word. The other between the second and third. If this were working correctly, these two badges would look different. If we were rendering multispace consistently on the badge and in text width computation, the characters wouldn't be splayed apart. That's a sign that the badge is too wide for the text.

https://img.shields.io/badge/tests-hello%20%20%20%20%20%20%20big%20world-blue.svg
https://img.shields.io/badge/tests-hello%20big%20%20%20%20%20%20%20world-blue.svg




If I'm right, this problem is for everyone, but just in case, here is a screenshot of what I'm seeing:

screen shot 2017-12-05 at 10 00 54 pm

A way to work around this is to use the half space character. Though, any fussing we do with spacing is kind of shot because the inaccurate text width.

**Comment by RedSparr0w** [comment]
We could replace the spaces with `%A0`

with `%20`
https://img.shields.io/badge/tests-hello%20%20%20%20%20%20%20big%20world-blue.svg

https://img.shields.io/badge/tests-hello%20big%20%20%20%20%20%20%20world-blue.svg

with `%A0`
https://img.shields.io/badge/tests-hello%A0%A0%A0%A0%A0%A0%A0big%A0world-blue.svg

https://img.shields.io/badge/tests-hello%A0big%A0%A0%A0%A0%A0%A0%A0world-blue.svg


Windows 10 & Chrome:
image

**Comment by paulmelnikow** [comment]
Widths look better!

Though when I tried to use `%A0` in this badge it caused problems with the symbols:

All spaces using `%A0`



All spaces using `%20`



```
https://img.shields.io/badge/tests-%E2%9C%94%A012%A0%E2%9C%98%A03%A0%E2%97%8F%A03-orange.svg
https://img.shields.io/badge/tests-%E2%9C%94%2012%20%E2%9C%98%203%20%E2%97%8F%203-orange.svg
```

**Comment by RedSparr0w** [comment]
Strange, no matter where in the url you change even just 1 of the `%20` to `%A0` it always produces the same result, even if you add it to the end

**Comment by RedSparr0w** [comment]
maybe somewhere we could replace all spaces with the `&nbsp;`|`&#160;` char as it shouldn't negatively affect other badges?
working example in `make-badge.js`:
```javascript
function escapeXml(s) {
  return s.replace(/&/g, '&amp;')
          .replace(/</g, '&lt;')
          .replace(/>/g, '&gt;')
          .replace(/"/g, '&quot;')
          .replace(/'/g, '&apos;')
          .replace(/ /g, '&#160;');
}
```

Not sure where would be best for it to go though.

**Comment by NZSmartie** [comment]
What about simple `/` separators?

[example](https://img.shields.io/badge/tests-%E2%9C%94%2012%20/%20%E2%9C%98%203%20/%20%E2%97%8F%203-orange.svg)
[example](https://img.shields.io/badge/tests-%E2%9C%94%2012%20%2F%20%E2%97%8F%203-green.svg)
[example](https://img.shields.io/badge/tests-%E2%9C%94%2012-brightgreen.svg)

Screenshot from Windows 10, Firefox:
image

**Comment by tooomm** [comment]
I really like horizontal lines:
 

Too bad macOS messes around with the fast-forward icon. I'm not convinced by the dot.
Also, macOS changes the checkmark to a black one which is bad as well.
Is there some other checkmark that works for apple and stays white?

By the way... would the badge turn red if it contains only fails, or fails+skipped?

**Comment by NZSmartie** [comment]
> @tooomm 
> By the way... would the badge turn red if it contains only fails, or fails+skipped?

yup, that's exactly what it does

**Comment by NZSmartie** [comment]
the reason why the symbols are inconsistent on each platform as the browser will fallback onto another font that supports that unicode range. A solution would be to include a second true type font that includes a standardised set of fonts. like Font Awesome for example

**Comment by NZSmartie** [comment]
How difficult would it be to embed a symbol (much like the brand logos) and use that for the ✔ ✘ and [skip symbol]?

**Comment by RedSparr0w** [comment]
For the sake of consistency i think the only way will be to embed an svg image for the check, cross, and other symbol

**Comment by paulmelnikow** [comment]
> Also, macOS changes the checkmark to a black one which is bad as well.

Where are you seeing that? I'm not seeing it in Chrome or Firefox.

example

I'm fine with that. ^^

It feels slightly cramped but we could tweak the spacing later on.

Another suggestion, riffing on the dot: midline horizontal ellipsis.

example

**Comment by NZSmartie** [comment]
@paulmelnikow have changed it to the full circle with horizontal bars. but the CI is failing due ot an error outside of my control. unless i need to merge with master?

**Comment by paulmelnikow** [comment]
Yea, merging master into this branch should fix it.

**Comment by RedSparr0w** [comment]
The checkmarks are black on iOS:
image
Also the extra squares aren't showing on this badge, but they are on the earlier example post you made:
image

**Comment by NZSmartie** [comment]
@RedSparr0w I think for now this will do and I'll see what I can do in regards to #127 / #552 to make it cleaner

**Comment by tooomm** [comment]
And/Or this:

>For the sake of consistency i think the only way will be to embed an svg image for the check, cross, and other symbol

By the way: font-awesome includes icons for github, twitter, telegram, bitbucket, facebook, gitlab, gratipay, slack etc. too!

**Comment by paulmelnikow** [comment]
After sleeping on it, I like rightward arrow to bar. Most of our suggestions have conveyed "maybe" or "indeterminate," but this one actually suggests skipping.

example

These are different checks and X's. Arguably the badge more legible as the symbols are lighter so the numbers stand out more. How do they look on iOS?

example

**Comment by RedSparr0w** [comment]
iOS 9 (safari):
image
Windows 10 (chrome):
image

**Comment by tooomm** [comment]
Is there some other thicker arrow available? This one is really hard to see at all!
Also, it doesn't need to point against a wall in my opinion because that symbolizes `end` or `to the end` more than `skip`.

The dot is kinda bold next to the others, that's why I searched http://graphemica.com/search?q=arrow...
(had to use the bolder `x`, couldn't find the other)











Also found a bit smaller dot:

**Comment by paulmelnikow** [comment]
Hi! Whoever picks this up, this work will need to be reworked a bit. A bunch of files moved in #1563, and an entirely rewritten copy of the appveyor service was merged in #1543.

**Comment by NZSmartie** [comment]
This issue became stalled as there was disagreement on what icons/symbols to use for representing the various results. If a consensus is made, i'll be happy to pick it up again

**Comment by kzu** [comment]
Plz, merge some form of this. Given users can customize the icons, the defaults aren't critical to agree on. 

I personally like this one mentioned above:

image

**Comment by paulmelnikow** [comment]
Reopened at #2000.

**Pull request closed without merging**